packages/perseus/src/util/parse-perseus-json/general-purpose-parsers/discriminated-union.test.ts

```diff
@@ -0,0 +1,80 @@
+import {discriminatedUnion} from "./discriminated-union";
+import {constant} from "./constant";
+import {object} from "./object";
+import {number} from "./number";
+import {parse} from "../parse";
+import {failure, success} from "../result";
+
+describe("a discriminatedUnion with one variant", () => {
+    const unionParser = discriminatedUnion(
+        object({type: constant("ok")}),
+        object({type: constant("ok"), value: number}),
+    ).parser;
+
+    it("parses a valid value", () => {
+        const input = {type: "ok", value: 3};
+
+        expect(parse(input, unionParser)).toEqual(success(input))
+    })
+
+    it("rejects a value with the wrong `type`", () => {
+        const input = {type: "bad", value: 3};
+
+        expect(parse(input, unionParser)).toEqual(
+            failure(`At (root).type -- expected "ok", but got "bad"`),
+        )
+    })
+
+    it("rejects a value with a valid type but wrong fields", () => {
+        const input = {type: "ok", value: "foobar"};
+
+        expect(parse(input, unionParser)).toEqual(
+            failure(`At (root).value -- expected number, but got "foobar"`),
```

Comment:
Would it be useful to include the discriminant info here?  Otherwise we don't know which union entry has bad data.  🤔  

```
At (root {type: "ok"}).value -- expected number, but got "footer"
```

Maybe that gets too verbose.

Reply:
I agree that would be better. The parser context currently doesn't have a way to track which variant of a union you're currently trying to parse, so I can't think of a way to add this info to the message easily.

I'll try adding a `forVariant` method to ParseContext, parallel to `forSubtree`. Maybe that will be easier than I'm imagining.